### PR TITLE
Simplify REST API nesting (#3323)

### PR DESCRIPTION
**Pull request by amiremohamadi**
/claim #3323

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by amiremohamadi** [comment]
there are two things that need to be considered:

1. for documentations, models are generated based on the code and internal structures. and since we used a `serde deserializer` for our purpose, the structure remained the same.

2. as @darsto [mentioned](https://github.com/qdrant/qdrant/issues/3323#issuecomment-1880077721) serde has some problems with parsing untagged enums and writing a simple deserializer wouldn't work perfectly (it breaks some of existing tests). with the help of `serde-untagged` (which provides helper `Visitors` for untagged enums) the problem is solved.
but if we don't want to add any dependencies, maybe with some efforts we will be able to implement our serde `Visitor` instead of using `serde-untagged`.

**Comment by darsto** [comment]
Putting all the implementation problems aside, the main concern is that JSON arrays could be already parsed as objects. This seems to be a standard JSON thing (at least every serde-json app works this way) so there's a chance someone uses this format.
- and we're breaking it now

**Comment by amiremohamadi** [comment]
@darsto i think we already solved it by using a different serde `Visitor`. can u provide some examples on how it can break backward compatibility?

**Comment by darsto** [comment]
I had to think a little but it seems you're right! :) I can't come up with anything that's broken now

File: lib/segment/src/types.rs [hunk 1]
```diff
@@ -2009,6 +2012,25 @@ impl Filter {
...
 
+fn deserialize_filter_item<'de, D>(deserializer: D) -> Result<Option<Vec<Condition>>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    use serde_untagged::UntaggedEnumVisitor;
+
+    #[derive(Deserialize)]
+    #[serde(untagged)]
+    enum Helper {
```
Comment (darsto) [inline lib/segment/src/types.rs:2023]: this seems unused

File: docs/redoc/master/openapi.json [hunk 2]
```diff
@@ -6187,6 +6187,7 @@
         "properties": {
           "should": {
             "description": "At least one of those conditions should match",
+            "default": null,
             "type": "array",
```
Comment (coszio) [inline docs/redoc/master/openapi.json:6191]: The changes in REST API should be reflected here as well

File: lib/segment/src/types.rs [hunk 3]
```diff
@@ -1954,12 +1954,15 @@ pub struct WithPayload {
...
     pub should: Option<Vec<Condition>>,
     /// All conditions must match
+    #[serde(default, deserialize_with = "deserialize_filter_item")]
     #[validate]
     pub must: Option<Vec<Condition>>,
     /// All conditions must NOT match
+    #[serde(default, deserialize_with = "deserialize_filter_item")]
     #[validate]
     pub must_not: Option<Vec<Condition>>,
 }
```
Comment (coszio) [inline lib/segment/src/types.rs:1968]: This works effectively, but it's hard to generate an openapi schema from this

**Review by coszio — CHANGES_REQUESTED** [review]
I like that it is only a matter of a few lines to achieve the deserialization like this, but we need to reflect it also in the OpenApi spec to propagate it to auto-generated clients. 

One alternative is using the untagged enum approach.

File: lib/segment/src/types.rs (hunk 3 above)
Comment (amiremohamadi) [reply lib/segment/src/types.rs:1968]: @coszio thanks for the review. although using untagged enums might be a clean way of fixing this, it requires a lot of changes in different parts of the code. so how about using `#[schemars(schema_with = "...")]` in order to generate the desired openapi schema for this?

Comment (coszio) [reply lib/segment/src/types.rs:1968]: Sure, let's see how that looks. In my opinion it looks like this parallel implementation might be harder to maintain. 

On another note, how about making the function generic and renaming to `deserialize_one_or_many`

**Comment by amiremohamadi** [comment]
conflicts are resolved. may I get a review on this?
@coszio 
@generall

**Review by generall — CHANGES_REQUESTED** [review]
I am not quite sure about the current version either. 
It creates a strange-looking OpenAPI "schema" 

image

Which is quite likely to be a problem for generated clients

**Comment by amiremohamadi** [comment]
> Which is quite likely to be a problem for generated clients

@generall  thanks for your review.

i think we used a similar schema for `SparseVectorParams` as well. however it's possible to generate sth like this, by some minor changes:



what's your opinion?

**Comment by amiremohamadi** [comment]
ping @generall @coszio

**Comment by coszio** [comment]
@amiremohamadi Unfortunately adding this `OneOrMany_for_Condition` will also be problematic with client generation. However, with a few tweaks your implementation can be fixed for a simpler schema.

I'll push them here, I hope you don't mind. I've also rebased this branch on top of current dev.

**Comment by coszio** [comment]
New look: 
image

**Review by generall — APPROVED** [review]

**Pull request merged**